### PR TITLE
Core - refactor

### LINKED ISSUE #306: Core - Core refactor
## Description

For clean up and better code structures
- remove `transaction_payload_type_t`, add essence enum when protocol support different types
- remove transaction wrapper methods in `core/payloads/transaction.c`
  - tx_payload_add_input
  - tx_payload_add_output
  - tx_payload_add_sig_block
  - tx_payload_add_ref_block

### PR DESCRIPTION
**Pull request by oopsmonk**
# Description of change

fixes #306 
- remove transaction_payload_type_t
- remove transaction wrapper methods
- unify the name of message payloads
- refactor marcors with `BIN_TO_XXX`
- `payload_len` stays in `transaction_essence_t` that we don't calculate it every time.
- rename `extended` output to `basic` output
- remove `core/seed.c/h`
- removed unused types from `types.h`

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

local tests are passed

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by matjazv — APPROVED** [review]

File: src/core/models/message.c [hunk 1]
```diff
@@ -221,7 +221,7 @@ size_t core_message_serialize(core_message_t* msg, byte_t buf[], size_t buf_len)
       payload_len = (uint32_t)tx_payload_serialize_length((transaction_payload_t*)msg->payload);
       break;
     case CORE_MESSAGE_PAYLOAD_TAGGED:
-      payload_len = (uint32_t)(uint32_t)tagged_data_serialize_len((tagged_data_t*)(msg->payload));
+      payload_len = (uint32_t)(uint32_t)tagged_data_serialize_len((tagged_data_payload_t*)(msg->payload));
```
Comment (rebinmannady93) [inline src/core/models/message.c]:
```suggestion
      payload_len = (uint32_t)tagged_data_serialize_len((tagged_data_payload_t*)(msg->payload));
```

File: src/core/models/outputs/output_basic.c [hunk 2]
```diff
@@ -61,13 +60,13 @@ output_extended_t* output_extended_new(uint64_t amount, native_tokens_t* tokens,
   }
 
   // create an extened output object
```
Comment (rebinmannady93) [inline src/core/models/outputs/output_basic.c]:
```suggestion
  // create a basic output object
```

**Review by rebinmannady93 — CHANGES_REQUESTED** [review]

File: src/core/models/outputs/output_basic.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 IOTA Stiftung
```
Comment (rebinmannady93) [inline src/core/models/outputs/output_basic.h:1]:
```suggestion
// Copyright 2022 IOTA Stiftung
```

File: src/client/api/json_parser/output_basic.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * @param[out] basic A new basic object
+ * @return int 0 on success
+ */
+int json_output_basic_deserialize(cJSON* output_obj, output_basic_t** basic);
+
+/**
+ * @brief Serialize basic output to JSON object
+ *
+ * @param basic An basic output
+ * @return cJSON* NULL on errors
```
Comment (rebinmannady93) [inline src/client/api/json_parser/output_basic.h:27]:
```suggestion
 * @return cJSON* or NULL on errors
```

**Review by rebinmannady93 — CHANGES_REQUESTED** [review]

File: src/core/models/message.c (hunk 1 above)
Comment (oopsmonk) [reply src/core/models/message.c]: thanks :)

File: src/client/api/json_parser/output_basic.h (hunk 4 above)
Comment (oopsmonk) [reply src/client/api/json_parser/output_basic.h:27]: `cJSON*` is not part of the description sentence.

File: src/core/models/outputs/output_basic.h (hunk 3 above)
Comment (oopsmonk) [reply src/core/models/outputs/output_basic.h:1]: renaming files will not bump the copyright year, no?

Comment (rebinmannady93) [reply src/core/models/outputs/output_basic.h:1]: Yes, you are right.

File: src/client/api/json_parser/output_basic.h (hunk 4 above)
Comment (rebinmannady93) [reply src/client/api/json_parser/output_basic.h:27]: I was referencing to below cases :
- [@return output_basic_t* or NULL on failure](https://github.com/iotaledger/iota.c/pull/339/files#diff-78a43ec0f14f3466dc2ec2c190be26668412bc20d60d46e7224171f87e371465R37)
- [@return output_basic_t* or NULL on failure](https://github.com/iotaledger/iota.c/pull/339/files#diff-78a43ec0f14f3466dc2ec2c190be26668412bc20d60d46e7224171f87e371465R72)
- [@return tagged_data_payload_t* or NULL on failure](https://github.com/iotaledger/iota.c/pull/339/files#diff-df9b6d3d491a8931542871c181747e15f60c8f18515da2e97b8024d6595d668bR77)

Comment (oopsmonk) [reply src/client/api/json_parser/output_basic.h:27]: documentation is not included in this PR, if you are happy I can fix the above sentences, what is your suggestion?

Comment (rebinmannady93) [reply src/client/api/json_parser/output_basic.h:27]: We can fix this while doing the documentation part.

**Review by rebinmannady93 — APPROVED** [review]

**Pull request merged**